Bohdan/feature/routes

BohdanSizomin opened the pull request.

BohdanSizomin committed:
fix

BohdanSizomin committed:
- poetry fixed
- settings file fixed
- Mail send implemented
- template for email created
- small models refactor
- login router extra logic added
 - extra schemas for pydantic (For creating users,for reseting password)
- tests coverage

BohdanSizomin committed:
migration added

BohdanSizomin committed:
- test error fixed

BohdanSizomin committed:
- test for user signing up and email sending created

BohdanSizomin committed:
all tests working for now

kipod commented on app/controller/mail.py:
```diff
@@ -0,0 +1,52 @@
+)
+
+
+fm = FastMail(conf)
```
fm -> mail

kipod replied:
```python
mail = FastMail(ConnectionConfig(**settings))
```

kipod commented on app/router/auth.py:
```diff
@@ -30,24 +33,29 @@ def login(
         user_credentials.username,
         user_credentials.password,
     )
+    if not user.is_verified:
```
add:
```python 
if not user:
   raise ...

kipod commented on app/router/auth.py:
```diff
@@ -30,24 +33,29 @@ def login(
     if not user:
-        log(log.INFO, "User [%s] does not exist \n", user_credentials.username)
+        log(log.WARNING, "User [%s] does not exist \n", user_credentials.username)
         raise HTTPException(status_code=403, detail="Invalid credentials")
```
move this code before line 41

kipod commented on app/router/auth.py:
```diff
@@ -61,20 +69,20 @@ def google_login(
             google_openid_key=user_data.google_openid_key,
         )
-        user = model.User(**user.dict())
+        user = m.User(**user.dict())
```
```python
user = m.User(
username=user_data.username,
            email=user_data.email,
            password=user_data.google_openid_key,
            google_openid_key=user_data.google_openid_key,
)
```

kipod commented on app/router/auth.py:
```diff
@@ -83,3 +91,15 @@ def google_login(
+    db.commit()
+    db.refresh(user)
+    await send_email(
+        user.email, settings.FRONTEND_RESET_PASSWORD_URL + user.verification_token
```
use: urljoin

kipod replied:
```python
# https://docs.python.org/3/library/urllib.parse.html
# https://stackoverflow.com/questions/70477787/how-to-get-current-path-in-fastapi-with-domain
await send_email(
        user.email, urllib.parse.urljoin(BASE_URL,  settings.FRONTEND_RESET_PASSWORD_URL, user.verification_token
)

kipod commented on app/router/auth.py:
```diff
@@ -83,3 +91,15 @@ def google_login(
+    await send_email(
+        user.email, settings.FRONTEND_RESET_PASSWORD_URL + user.verification_token
+    )
+    return HTTPStatus.OK
```
add ```try-except```

kipod commented on app/router/user.py:
```diff
@@ -12,35 +15,25 @@
 
 @router.post("/", status_code=201, response_model=schema.UserOut)
 def create_user(user: schema.UserCreate, db: Session = Depends(get_db)):
```
Consider to remove this route

kipod commented on app/router/user.py:
```diff
@@ -77,11 +71,47 @@ def get_user(
+        raise HTTPException(status_code=HTTPStatus.NOT_FOUND, detail="User not found")
+
+    user.password = data.password
+    user.verification_token = None
```
user.verification_token = ""

kipod commented on app/schema/email.py:
```diff
@@ -0,0 +1,7 @@
+from pydantic import EmailStr, BaseModel
+
+
+class EmailSchema(BaseModel):
```
???

kipod commented on app/schema/token.py:
```diff
@@ -8,4 +7,4 @@ class Token(BaseModel):
 
 class TokenData(BaseModel):
-    id: Optional[str] = None
+    id: int
```
id -> user_id

kipod commented on app/templates/email_template.html:
```diff
@@ -0,0 +1,466 @@
+                      margin: 0;
+                      padding: 0 16px;
+                    ">
+                                    <!--[if (gte mso 9)|(IE)]>
+                      <table align="center" role="presentation">
+                        <tbody>
+                          <tr>
+                            <td width="600">
+                    <![endif]-->
```
remove commented lines!!!

kipod commented on app/templates/email_template.html:
```diff
@@ -0,0 +1,466 @@
+                                                line-height: 33.6px;
+                                                margin: 0;
+                                              " align="left">
+                                                                                        Hi {{ email }},
```
use username!

kipod commented on tests/test_auth.py:
```diff
@@ -0,0 +1,56 @@
+        username=USER_NAME,
+        google_openid_key=USER_GOOGLE_ID,
+    )
+
```
Pls add comments!!

kipod commented on tests/test_auth.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def test_email_password_reset(client: TestClient, db: Session):
+    fm.config.SUPPRESS_SEND = 1
```
move this in test fixture ```client```

kipod commented on tests/test_auth.py:
```diff
@@ -0,0 +1,56 @@
+
+def test_email_password_reset(client: TestClient, db: Session):
+    fm.config.SUPPRESS_SEND = 1
+    request = schema.UserSignUp(email=USER_EMAIL, username=USER_NAME)
```
request --> request_data

kipod commented on tests/test_auth.py:
```diff
@@ -0,0 +1,56 @@
+    user = db.query(model.User).filter_by(email=USER_EMAIL).first()
+    assert not user.is_verified
+    assert user.verification_token
+    assert response.status_code == 200
```
add ``` client.post("api/reset_password", json=s.PasswordResetData(verification_token=..., password=...))```

kipod commented on tests/test_user.py:
```diff
@@ -83,14 +88,29 @@ def test_auth(client: TestClient, db: Session):
+
+    user = db.query(model.User).filter_by(email=USER_EMAIL).first()
 
-    response = client.post("/api/google_login", json=request.dict())
-    assert response and response.ok, "unexpected response"
+    assert user
+    assert not user.is_verified
 
-    token = schema.Token.parse_obj(response.json())
-    headers = {"Authorization": f"Bearer {token.access_token}"}
-    assert headers
+    data_reset_password = schema.ResetPasswordData(
+        password=USER_PASSWORD, verification_token=user.verification_token
+    )
+    response = client.post("api/user/reset_password", json=data_reset_password.dict())
+
+    assert response.ok
```
pls move this code into ```test_auth()```

kipod reviewed (changes requested):
Thank you.
Please fix some issues....

BohdanSizomin committed:
- username is not needed to login
- User model and routes cleanup
- tests logic been changed due to cleanup
- migrations
